core/modules/loader/WWorkerList.cc

```diff
@@ -0,0 +1,406 @@
+                                     " res=" << res.second);
+        }
+    }
+
+
+    // TODO probably special action should be taken if this is our name.
+    LOGS(_log, LOG_LVL_INFO, "updateEntry strRange=" << strRange);
+    if (strRange.getValid()) {
+        // Does the new range match the old range?
+        auto oldRange = item->setRangeStr(strRange);
+        LOGS(_log, LOG_LVL_INFO, "updateEntry set name=" << wId << " range=" << strRange);
+        if (not oldRange.equal(strRange)) {
+            // Since the value changed, it needs to be removed and reinserted.
+            // No invalid ranges should be in the map.
+            if (oldRange.getValid()) {
+                // The old value was valid, so it is likely in the map.
+                auto rangeIter = _rangeMap.find(oldRange);
+                if (rangeIter != _rangeMap.end()) {
+                    _rangeMap.erase(rangeIter);
+                }
+            }
+            if (strRange.getValid()) {
+                _rangeMap[strRange] = item;
+            }
+        }
+    }
+}
+
+
+WWorkerListItem::Ptr WWorkerList::findWorkerForKey(std::string const& key) {
+    std::string const funcName("WWorkerList::findWorkerForKey");
+    std::unique_lock<std::mutex> lk(_mapMtx);
+    // TODO Really could use a custom container for _rangeMap to speed this up.
+    for (auto const& elem : _rangeMap) {
+        if (elem.second->containsKey(key)) {
+            LOGS(_log, LOG_LVL_INFO, funcName << " key=" << elem.first << " -> " << *(elem.second));
+            return elem.second;
+        }
+    }
+    LOGS(_log, LOG_LVL_WARN, funcName << " did not find worker for key=" << key);
+    return nullptr;
+}
+
+
+
+void WWorkerListItem::addDoListItems(Central *central) {
+    std::lock_guard<std::mutex> lck(_mtx);
+    if (_workerUpdateNeedsMasterData == nullptr) {
+        _workerUpdateNeedsMasterData.reset(new WorkerNeedsMasterData(shared_from_this(), _central));
+        central->addDoListItem(_workerUpdateNeedsMasterData);
+    }
+}
+
+
+bool WWorkerListItem::equal(WWorkerListItem& other) {
+    std::string const funcName("WWorkerListItem::equal");
+    std::lock(_mtx, other._mtx);
+    std::lock_guard<std::mutex> lck1(_mtx,       std::adopt_lock);
+    std::lock_guard<std::mutex> lck2(other._mtx, std::adopt_lock);
+    if (_name != other._name) {
+        LOGS(_log, LOG_LVL_INFO, funcName << " item name not equal t=" << _name << " o=" << other._name);
+        return false;
+    }
+    if (*_udpAddress != *other._udpAddress) {
+        LOGS(_log, LOG_LVL_INFO, funcName << " item addr != name=" << _name <<
+                                 " t=" << *_udpAddress << " o=" << *other._udpAddress);
+        return false;
+    }
+
+    if (not _range.equal(other._range)) {
+        LOGS(_log, LOG_LVL_INFO, funcName << " item range != name=" << _name <<
+                                 " t=" << _range << " o=" << other._range);
```

Comment:
this seems to have similar locking issues as MWorkerList, and if those are valid issues they need to be fixed here too.
Perhaps there's an opportunity to move some duplicated code/functionality into a base class.

Reply:
point of order: this comment ended up on the wrong line. It was intended for the ostream operator, which probably should be taking a lock before streaming out variables like `_range`